pupil_src/shared_modules/surface_tracker/surface_tracker_online.py

```diff
@@ -0,0 +1,154 @@
+"""
+(*)~---------------------------------------------------------------------------
+Pupil - eye tracking platform
+Copyright (C) 2012-2018 Pupil Labs
+
+Distributed under the terms of the GNU
+Lesser General Public License (LGPL v3.0).
+See COPYING and COPYING.LESSER for license details.
+---------------------------------------------------------------------------~(*)
+"""
+
+import logging
+
+logger = logging.getLogger(__name__)
+
+import pyglui
+import pyglui.cygl.utils as pyglui_utils
+import gl_utils
+
+
+from surface_tracker import Heatmap_Mode
+from surface_tracker.surface_tracker import Surface_Tracker
+from surface_tracker.surface_online import Surface_Online
+
+
+class Surface_Tracker_Online(Surface_Tracker):
+    def __init__(self, g_pool, marker_min_perimeter=60, inverted_markers=False):
+        self.Surface_Class = Surface_Online
+        self.freeze_scene = False
+        self.frozen_scene_frame = None
+        self.frozen_scene_tex = None
+        super().__init__(
+            g_pool,
+            marker_min_perimeter=marker_min_perimeter,
+            inverted_markers=inverted_markers,
+        )
+
+        self.ui_info_text = "This plugin detects and tracks fiducial markers visible in the scene. You can define surfaces using 1 or more marker visible within the world view by clicking *add surface*. You can edit defined surfaces by selecting *Surface edit mode*."
+        self.supported_heatmap_modes = [Heatmap_Mode.WITHIN_SURFACE]
+
+        self.menu = None
+        self.button = None
+        self.add_button = None
+
+    @property
+    def _save_dir(self):
+        return self.g_pool.user_dir
+
+    @property
+    def has_freeze_feature(self):
+        return True
+
+    def _update_ui_custom(self):
+        def set_freeze_scene(val):
+            self.freeze_scene = val
+            if val:
+                self.frozen_scene_tex = pyglui_utils.Named_Texture()
+                self.frozen_scene_tex.update_from_ndarray(self.current_frame.img)
+            else:
+                self.frozen_scene_tex = None
+
+        self.menu.append(
+            pyglui.ui.Switch(
+                "freeze_scene", self, label="Freeze Scene", setter=set_freeze_scene
+            )
+        )
+
+    def _per_surface_ui_custom(self, surface, s_menu):
```

Comment:
I would prefer `surf_menu`

Reply:
You mean `_per_surface_ui_custom --> surf_menu` ? This would be a bit ambiguous with `_per_surface_ui`.

Reply:
oh sorry, I meant instead of `s_menu`